File: benchmarks/Java/src/richards/Scheduler.java
/*
 * This benchmark is derived from Mario Wolczko's Java and Smalltalk version of
 * Richards.
 *
 * It is modified based on the SOM version and to use Java 8 features.
 * License details:
 *   http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html
 */
package richards;

import java.util.Arrays;

public class Scheduler extends RBObject {
  private TaskControlBlock taskList;
  private TaskControlBlock currentTask;
  private int currentTaskIdentity;
  private final TaskControlBlock[] taskTable;

  private int queuePacketCount;
  private int holdCount;

  private int layout;

  private static final boolean TRACING = false;

  public Scheduler() {
    // init tracing
    layout  = 0;

    // init scheduler
    queuePacketCount = 0;
    holdCount = 0;
    taskTable = new TaskControlBlock[NUM_TYPES];
    Arrays.fill(taskTable, NO_TASK);
    taskList = NO_TASK;
  }


  void createDevice(final int identity, final int priority,
      final Packet workPacket, final TaskState state) {
    DeviceTaskDataRecord data = new DeviceTaskDataRecord();

    createTask(identity, priority, workPacket, state,
       (final Packet workArg, final RBObject wordArg) -> {
         DeviceTaskDataRecord dataRecord = (DeviceTaskDataRecord) wordArg;
         Packet functionWork = workArg;
         if (NO_WORK == functionWork) {
           if (NO_WORK == (functionWork = dataRecord.getPending())) {
             return markWaiting();
           } else {
             dataRecord.setPending(NO_WORK);
             return queuePacket(functionWork);
           }
         } else {
           dataRecord.setPending(functionWork);
           if (TRACING) {
             trace(functionWork.getDatum());
           }
           return holdSelf();
         }},
      data);
  }

  void createHandler(final int identity, final int priority,
      final Packet workPaket, final TaskState state) {
    HandlerTaskDataRecord data = new HandlerTaskDataRecord();
    createTask(identity, priority, workPaket, state,
        (final Packet work, final RBObject word) -> {
          HandlerTaskDataRecord dataRecord = (HandlerTaskDataRecord) word;
          if (NO_WORK != work) {
            if (WORK_PACKET_KIND == work.getKind()) {
              dataRecord.workInAdd(work);
            } else {
              dataRecord.deviceInAdd(work);
            }
          }

          Packet workPacket;
          if (NO_WORK == (workPacket = dataRecord.workIn())) {
            return markWaiting();
          } else {
            int count = workPacket.getDatum();
            if (count >= Packet.DATA_SIZE) {
              dataRecord.workIn(workPacket.getLink());
              return queuePacket(workPacket);
            } else {
              Packet devicePacket;
              if (NO_WORK == (devicePacket = dataRecord.deviceIn())) {
                return markWaiting();
              } else {
                dataRecord.deviceIn(devicePacket.getLink());
                devicePacket.setDatum(workPacket.getData()[count]);
                workPacket.setDatum(count + 1);
                return queuePacket(devicePacket);
              }
            }
          }
      }, data);
  }

  void createIdler(final int identity, final int priority, final Packet work,
      final TaskState state) {
        IdleTaskDataRecord data = new IdleTaskDataRecord();
        createTask(identity, priority, work, state,
            (final Packet workArg, final RBObject wordArg) -> {
              IdleTaskDataRecord dataRecord = (IdleTaskDataRecord) wordArg;
              dataRecord.setCount(dataRecord.getCount() - 1);
              if (0 == dataRecord.getCount()) {
                return holdSelf();
              } else {
                if (0 == (dataRecord.getControl() & 1)) {
                  dataRecord.setControl(dataRecord.getControl() / 2);
                  return release(DEVICE_A);
                } else {
                  dataRecord.setControl((dataRecord.getControl() / 2) ^ 53256);
                  return release(DEVICE_B);
                }
              }
            }, data);
  }

  Packet createPacket(final Packet link, final int identity, final int kind) {
    return new Packet(link, identity, kind);
  }

  void createTask(final int identity, final int priority,
      final Packet work, final TaskState state,
      final ProcessFunction aBlock, final RBObject data) {

    TaskControlBlock t = new TaskControlBlock(taskList, identity,
        priority, work, state, aBlock, data);
    taskList = t;
    taskTable[identity] = t;
  }

  void createWorker(final int identity, final int priority,
      final Packet workPaket, final TaskState state) {
    WorkerTaskDataRecord dataRecord = new WorkerTaskDataRecord();
    createTask(identity, priority, workPaket, state,

        (final Packet work, final RBObject word) -> {
          WorkerTaskDataRecord data = (WorkerTaskDataRecord) word;
          if (NO_WORK == work) {
            return markWaiting();
          } else {
            data.setDestination((HANDLER_A == data.getDestination()) ? HANDLER_B : HANDLER_A);
            work.setIdentity(data.getDestination());
            work.setDatum(0);
            for (int i = 0; i < Packet.DATA_SIZE; i++) {
              data.setCount(data.getCount() + 1);
              if (data.getCount() > 26) { data.setCount(1); }
              work.getData()[i] = 65 + data.getCount() - 1;
            }
            return queuePacket(work);
          }
        }, dataRecord);
  }

  public boolean start() {
    Packet workQ;

    createIdler(IDLER, 0, NO_WORK, TaskState.createRunning());
    workQ = createPacket(NO_WORK, WORKER, WORK_PACKET_KIND);
    workQ = createPacket(workQ,   WORKER, WORK_PACKET_KIND);

    createWorker(WORKER, 1000, workQ, TaskState.createWaitingWithPacket());
    workQ = createPacket(NO_WORK, DEVICE_A, DEVICE_PACKET_KIND);
    workQ = createPacket(workQ,   DEVICE_A, DEVICE_PACKET_KIND);
    workQ = createPacket(workQ,   DEVICE_A, DEVICE_PACKET_KIND);

    createHandler(HANDLER_A, 2000, workQ, TaskState.createWaitingWithPacket());
    workQ = createPacket(NO_WORK, DEVICE_B, DEVICE_PACKET_KIND);
    workQ = createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);
    workQ = createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);

    createHandler(HANDLER_B, 3000, workQ, TaskState.createWaitingWithPacket());
    createDevice(DEVICE_A, 4000, NO_WORK, TaskState.createWaiting());
    createDevice(DEVICE_B, 5000, NO_WORK, TaskState.createWaiting());

    schedule();

    return queuePacketCount == 23246 && holdCount == 9297;
  }

  TaskControlBlock findTask(final int identity) {
    TaskControlBlock t = taskTable[identity];
    if (NO_TASK == t) { throw new RuntimeException("findTask failed"); }
    return t;
  }

  TaskControlBlock holdSelf() {
    holdCount = holdCount + 1;
    currentTask.setTaskHolding(true);
    return currentTask.getLink();
  }

  TaskControlBlock queuePacket(final Packet packet) {
    TaskControlBlock t = findTask(packet.getIdentity());
    if (NO_TASK == t) { return NO_TASK; }

    queuePacketCount = queuePacketCount + 1;

    packet.setLink(NO_WORK);
    packet.setIdentity(currentTaskIdentity);
    return t.addInputAndCheckPriority(packet, currentTask);
  }

  TaskControlBlock release(final int identity) {
    TaskControlBlock t = findTask(identity);
    if (NO_TASK == t) { return NO_TASK; }
    t.setTaskHolding(false);
    if (t.getPriority() > currentTask.getPriority()) {
      return t;
    } else {
      return currentTask;
    }
  }

  void trace(final int id) {
    layout = layout - 1;
    if (0 >= layout) {
      // Checkstyle: stop
      System.out.println();
      // Checkstyle: resume
      layout = 50;
    }
    // Checkstyle: stop
    System.out.print(id);
    // Checkstyle: resume
  }

  TaskControlBlock markWaiting() {
    currentTask.setTaskWaiting(true);
    return currentTask;
  }

  void schedule() {
    currentTask = taskList;
    while (NO_TASK != currentTask) {
      if (currentTask.isTaskHoldingOrWaiting()) {
        currentTask = currentTask.getLink();
      } else {
        currentTaskIdentity = currentTask.getIdentity();
        if (TRACING) { trace(currentTaskIdentity); }
        currentTask = currentTask.runTask();
      }
    }
  }
}

File: benchmarks/C++/src/richards/Scheduler.cpp
#include "Scheduler.h"

namespace richards {
    Scheduler::Scheduler() {
        RBObject::initializeConstants();
        // init tracing
        _layout  = 0;

        // init scheduler
        _queuePacketCount = 0;
        _holdCount = 0;
        _taskTable = new shared_ptr<TaskControlBlock>[NUM_TYPES];

        for (int i = 0; i < NUM_TYPES; i++) {
            _taskTable[i] = RBObject::NO_TASK;
        }
        _taskList = RBObject::NO_TASK;
    }


    void Scheduler::createDevice(int identity, int priority, shared_ptr<Packet> workPacket, shared_ptr<TaskState> state) {
        shared_ptr<DeviceTaskDataRecord> data = make_shared<DeviceTaskDataRecord>();

        createTask(identity, priority, workPacket, state,
                    [&](shared_ptr<Packet> workArg, shared_ptr<RBObject> wordArg) -> shared_ptr<TaskControlBlock> {
        shared_ptr<DeviceTaskDataRecord> dataRecord = dynamic_pointer_cast<DeviceTaskDataRecord>(wordArg);
        shared_ptr<Packet> functionWork = workArg;
        if (RBObject::NO_WORK == functionWork) {
            if (RBObject::NO_WORK == (functionWork = dataRecord->getPending())) {
                return markWaiting();
            } else {
                dataRecord->setPending(RBObject::NO_WORK);
                return queuePacket(functionWork);
            }
        } else {
            dataRecord->setPending(functionWork);
                if (TRACING) {
                    trace(functionWork->getDatum());
                }
                return holdSelf();
            }
        }, data);
    }

    void Scheduler::createHandler(int identity, int priority,
        shared_ptr<Packet> workPaket, shared_ptr<TaskState> state) {

        shared_ptr<HandlerTaskDataRecord> data = make_shared<HandlerTaskDataRecord>();

        createTask(identity, priority, workPaket, state,
                    [&](shared_ptr<Packet> work, shared_ptr<RBObject> word) -> shared_ptr<TaskControlBlock> {
            shared_ptr<HandlerTaskDataRecord> dataRecord = dynamic_pointer_cast<HandlerTaskDataRecord>(word);
            if (RBObject::NO_WORK != work) {
                if (WORK_PACKET_KIND == work->getKind()) {
                dataRecord->workInAdd(work);
                } else {
                dataRecord->deviceInAdd(work);
                }
            }

            shared_ptr<Packet> workPacket;
            if (RBObject::NO_WORK == (workPacket = dataRecord->workIn())) {
                return markWaiting();
            } else {
                int count = workPacket->getDatum();
                if (count >= Packet::DATA_SIZE) {
                    dataRecord->workIn(workPacket->getLink());
                    return queuePacket(workPacket);
                } else {
                    shared_ptr<Packet> devicePacket;
                    if (RBObject::NO_WORK == (devicePacket = dataRecord->deviceIn())) {
                        return markWaiting();
                    } else {
                        dataRecord->deviceIn(devicePacket->getLink());
                        devicePacket->setDatum(workPacket->getData()[count]);
                        workPacket->setDatum(count + 1);
                        return queuePacket(devicePacket);
                    }
                }
            }
        }, data);
    }

    void Scheduler::createIdler(int identity, int priority, shared_ptr<Packet> work,
        shared_ptr<TaskState> state) {

        shared_ptr<IdleTaskDataRecord> data = make_shared<IdleTaskDataRecord>();

        createTask(identity, priority, work, state,
                    [&](shared_ptr<Packet> workArg, shared_ptr<RBObject> wordArg) -> shared_ptr<TaskControlBlock> {
            shared_ptr<IdleTaskDataRecord> dataRecord = dynamic_pointer_cast<IdleTaskDataRecord>(wordArg);
            dataRecord->setCount(dataRecord->getCount() - 1);
            if (0 == dataRecord->getCount()) {
                return holdSelf();
            } else {
                if (0 == (dataRecord->getControl() & 1)) {
                    dataRecord->setControl(dataRecord->getControl() / 2);
                    return release(DEVICE_A);
                } else {
                    dataRecord->setControl((dataRecord->getControl() / 2) ^ 53256);
                    return release(DEVICE_B);
                }
            }
        }, data);
    }

    shared_ptr<Packet> Scheduler::createPacket(shared_ptr<Packet> link, int identity, int kind) {
        return make_shared<Packet>(link, identity, kind);
    }

    void Scheduler::createTask(int identity, int priority,
        shared_ptr<Packet> work, shared_ptr<TaskState> state,
        function<shared_ptr<TaskControlBlock>(shared_ptr<Packet> work, shared_ptr<RBObject> word)> aBlock,
        shared_ptr<RBObject> data) {

        shared_ptr<TaskControlBlock> t = make_shared<TaskControlBlock>(_taskList, identity,
            priority, work, state, aBlock, data);
        _taskList = t;
        _taskTable[identity] = t;
    }

    void Scheduler::createWorker(int identity, int priority,
        shared_ptr<Packet> workPaket, shared_ptr<TaskState> state) {

        shared_ptr<WorkerTaskDataRecord> dataRecord = make_shared<WorkerTaskDataRecord>();

        createTask(identity, priority, workPaket, state,
            [&](shared_ptr<Packet> work, shared_ptr<RBObject> word) -> shared_ptr<TaskControlBlock> {
            shared_ptr<WorkerTaskDataRecord> data = dynamic_pointer_cast<WorkerTaskDataRecord>(word);
            if (RBObject::NO_WORK == work) {
                return markWaiting();
            } else {
                data->setDestination((HANDLER_A == data->getDestination()) ? HANDLER_B : HANDLER_A);
                work->setIdentity(data->getDestination());
                work->setDatum(0);
                for (int i = 0; i < Packet::DATA_SIZE; i++) {
                    data->setCount(data->getCount() + 1);
                    if (data->getCount() > 26) { 
                        data->setCount(1); 
                    }
                    work->getData()[i] = 65 + data->getCount() - 1;
                }
                return queuePacket(work);
            }
        }, dataRecord);
    }

    bool Scheduler::start() {
        shared_ptr<Packet> workQ;
        createIdler(IDLER, 0, RBObject::NO_WORK, TaskState::createRunning());
        workQ = createPacket(RBObject::NO_WORK, WORKER, WORK_PACKET_KIND);
        workQ = createPacket(workQ,   WORKER, WORK_PACKET_KIND);

        createWorker(WORKER, 1000, workQ, TaskState::createWaitingWithPacket());
        workQ = createPacket(RBObject::NO_WORK, DEVICE_A, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ,   DEVICE_A, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ,   DEVICE_A, DEVICE_PACKET_KIND);

        createHandler(HANDLER_A, 2000, workQ, TaskState::createWaitingWithPacket());
        workQ = createPacket(RBObject::NO_WORK, DEVICE_B, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);

        createHandler(HANDLER_B, 3000, workQ, TaskState::createWaitingWithPacket());
        createDevice(DEVICE_A, 4000, RBObject::NO_WORK, TaskState::createWaiting());
        createDevice(DEVICE_B, 5000, RBObject::NO_WORK, TaskState::createWaiting());

        schedule();

        return _queuePacketCount == 23246 && _holdCount == 9297;
    }

    shared_ptr<TaskControlBlock> Scheduler::findTask(int identity) {
        shared_ptr<TaskControlBlock> t = _taskTable[identity];
        if (RBObject::NO_TASK == t) { 
            throw Error("findTask failed");
        }
        return t;
    }

    shared_ptr<TaskControlBlock> Scheduler::holdSelf() {
        _holdCount = _holdCount + 1;
        _currentTask->setTaskHolding(true);
        return _currentTask->getLink();
    }

    shared_ptr<TaskControlBlock> Scheduler::queuePacket(shared_ptr<Packet> packet) {
        shared_ptr<TaskControlBlock> t = findTask(packet->getIdentity());
        if (RBObject::NO_TASK == t) { 
            return RBObject::NO_TASK; 
        }

        _queuePacketCount = _queuePacketCount + 1;

        packet->setLink(RBObject::NO_WORK);
        packet->setIdentity(_currentTaskIdentity);
        return t->addInputAndCheckPriority(packet, _currentTask);
    }

    shared_ptr<TaskControlBlock> Scheduler::release(int identity) {
        shared_ptr<TaskControlBlock> t = findTask(identity);
        if (RBObject::NO_TASK == t) { 
            return RBObject::NO_TASK; 
        }
        t->setTaskHolding(false);
        if (t->getPriority() > _currentTask->getPriority()) {
            return t;
        } else {
            return _currentTask;
        }
    }

    void Scheduler::trace(int id) {
        _layout = _layout - 1;
        if (0 >= _layout) {
            cout << endl;
            _layout = 50;
        }
        cout << id << endl;
    }

    shared_ptr<TaskControlBlock> Scheduler::markWaiting() {
        _currentTask->setTaskWaiting(true);
        return _currentTask;
    }

    void Scheduler::schedule() {
        _currentTask = _taskList;
        while (RBObject::NO_TASK != _currentTask) {
            if (_currentTask->isTaskHoldingOrWaiting()) {
                _currentTask = _currentTask->getLink();
            } else {
                _currentTaskIdentity = _currentTask->getIdentity();
                if (TRACING) {
                    trace(_currentTaskIdentity);
                }
                _currentTask = _currentTask->runTask();
            }
        }
    }
}